docs/type-annotations.org

```diff
@@ -0,0 +1,219 @@
+be returned and so the type actually must be the same.  The =(elsa
+(mixed) mixed)= signature allows us to pass in an =int= and it can return
+back a =string= no problem and so it would not catch a huge number of
+possible errors.
+
+What we want to express here is "X comes in, X comes out".
+
+*** Syntax
+
+The syntax for generic types is "generic type name" + =*= suffix.  Any
+string can be used for the generic type name, but customarily
+single-letter names are used.
+
+For the above mentioned identity function we therefore write the type
+as =(elsa (a*) a*)= where =a*= stands for a generic type =a=.
+
+A function such as =car= can be typed as follows:
+
+#+BEGIN_SRC elisp
+(elsa ((cons a* b*)) a*)
+#+END_SRC
+
+It takes a cons with =a= in the =car= and =b= in the =cdr= and return the =car=
+which is of type =a= , whatever that happens to be.
+
+** Optional types
+
+If a function can take optional arguments we preceed them with
+=&optional= marker just as we do in the argument list.
+
+#+BEGIN_SRC elisp
+(defun drop-items (list &optional n)
+  "Drop first item of LIST or N items if N is provided."
+  (declare (elsa ((list a*) &optional int) (list a*)))
+  (setq n (or n 1))
+  (dotimes (_ n list)
+    (setq list (cdr list))))
+#+END_SRC
+
+** Variadic types
+
+If a function can take arbitrary number of arguments we preceed the
+last variadic argument with =&rest= marker just as we do in the argument
+list.
+
+#+BEGIN_SRC elisp
+(defun join (separator &rest strings)
+  "Join STRINGS with SEPARATOR."
+  (declare (elsa (string &rest string) string))
+  (mapconcat 'identity strings separator))
+#+END_SRC
+
+** Type combinators
+*** Sum types
+:PROPERTIES:
+:ID:       5a21a68a-4df1-4d44-a854-1d9700858a1a
+:END:
+
+Sum types can be specified as a list form starting with =or=, so =(or
+string int)= is a type accepting strings or integers.
+
+A sum type is useful if the function internally checks the passed
+value and decides what processing to do:
+
+*TODO:* I think the double parens in the "argument" portion will lead to
+a lot of confusion.  People might expect to write just =(elsa (or int
+string) int)=.  I'm not sure if we should omit the "argument wrapping"
+parens or not.
+
+#+BEGIN_SRC elisp
+(defun to-number (x)
+  (declare (elsa ((or int string)) int))
+  (cond
+   ((numberp x) x)
+   ((stringp x) (string-to-number x))))
+#+END_SRC
+
+*** Intersection types
+
+Intersection types can be specified as list form starting with =and=, so
+=(and string float)= is a type which is at the same time string and
+float (such a type has empty domain, nothing can be string and float
+at the same time).  Intersection types are used to track impossible
+assignments.
+
+#+BEGIN_SRC elisp
+;; Such a condition can never evaluate to true
+(if (and (stringp x) (integerp x))
+    "Is both string and int"
+  "This is always true")
+#+END_SRC
+
+*** Difference types
+
+Difference types can be specified as list form starting with =-= so =(-
```

Comment:
It's kind of odd to have `and` and `or`, but `-` for this one. How about something like `except`?

Reply:
Yea, I wasn't sure what to use. My intention was to use an actual existing function to make it more familiar and memorable.  There's `cl-set-difference` of course, but... :D

Reply:
I'm going with `diff` for now.